# 데이터와 행렬
+ Linear Algebra는 데이터 분석에 필요한 각종 계산을 돕는다.
+ Linear Algebra는 대량의 데이터를 포함하는 복잡한 계산 과정을 간단한 수식으로 표현할 수 있다.

In [1]:
import numpy as np
import matplotlib.pyplot as plt

## 데이터의 유형

+ 선형대수에서 다루는 데이터는 형태에 따라 다음과 같이 나뉜다.
    + **`Scalar`** : 하나의 **숫자**
    + **`Vector`** : 여러 숫자로 이루어진 **데이터 레코드**
    + **`Matrix`** : **Data Record의 집합**
    + **`Tensor`** : 같은 크기의 **Matrix의 집합**
+ 

In [2]:
from sklearn.datasets import load_iris

iris = load_iris()
iris.data[0, :]

array([5.1, 3.5, 1.4, 0.2])

### 스칼라
+ 스칼라는 하나의 숫자만으로 이루어진 데이터(값)이다.
+ 스칼라는 보통 $ x $와 같이 알파벳 소문자로 표기하며
+ 실수인 숫자 중 하나이므로 실수 집합 $ R $의 원소로 표기한다.
    + $ x \in R $

In [9]:
feature_name_0 = iris.feature_names[0]
feature_scalar = iris.data[0][0]

print('IRIS 데이터의 0번째 Data Record의 0번째 값(Scalar)')
print(feature_name_0,':',feature_scalar)

IRIS 데이터의 0번째 Data Record의 0번째 값(Scalar)
sepal length (cm) : 5.1


### 벡터
+ 벡터는 여러개의 숫자가 특정한 순서대로 모여있는 것이다.
+ 여러개의 데이터를 따로따로 다루기 보다는 하나의 묶음(tuple)로 다룬다.
+ 이러한 값의 묶음을 Linear Algebra에서는 `Vector`라고 부른다.

+ 붓꽃 데이터는 4개의 데이터 $ (x_1, x_2, x_3, x_4) $가 하나로 묶여 있는데,
+ 이를 선형대수 기호로는 다음처럼 하나의 문자 $ x $로 표기한다.
    + $ \begin{bmatrix} x_1 \\ x_2 \\ x_3 \\ x_4 \\ \end{bmatrix} $
    + 또는, $ x \in R^4 $
+ 하나의 벡터를 이루는 데이터의 수가 n개이면 이를 n-dimensional vector라고 한다.
    + $ \begin{bmatrix} x_1 \\ x_2 \\ \vdots \\ x_N \\ \end{bmatrix} $
    + 또는, $ x \in R^N $ 

In [11]:
vector_0 = iris.data[0, :]
vector_0

array([5.1, 3.5, 1.4, 0.2])

$ vector_0 = \begin{bmatrix} 5.1 \\ 3.5 \\ 1.4 \\ 0.2 \end{bmatrix} $

<br><br>
+ `x1 = np.array([[5.1], [3.5], [1.4], [0.2]])`
+ 위 처럼 2차원 배열로 표기하지만,
+ numpy에서는 1차원 배열 객체를 대부분 벡터로 인정한다.
    + 이때는 벡터가 **하나의 행처럼 표시되어도 실제로는 열이라는 점에 주의**한다.

### 행렬
+ 행렬의 복수의 차원을 가지는 Data Record가 여러개 있는 경우를 표기한 것이다.
+ 4차원의 Data Record가 6개 있다면 아래와 같은 행렬로 표현 할 수 있다.
    + $ X = \begin{bmatrix} x_{1,1} & x_{1,2} & x_{1,3} & x_{1,4} \\
    x_{2,1} & x_{2,2} & x_{2,3} & x_{2,4} \\
    x_{3,1} & x_{3,2} & x_{3,3} & x_{3,4} \\
    x_{4,1} & x_{4,2} & x_{4,3} & x_{4,4} \\
    x_{5,1} & x_{5,2} & x_{5,3} & x_{5,4} \\
    x_{6,1} & x_{6,2} & x_{6,3} & x_{6,4} \end{bmatrix} $